### PR TITLE
[Quantization] enable trt calibration

### PR DESCRIPTION
**Pull request by chenbohua3**

### CODE REVIEW

File: pytorch_blade/torch_blade/mlir/disc_engine_conversion.py [hunk 1]
```diff
@@ -105,7 +105,9 @@ def _get_mlir_unsupported(graph):
     return unspt_nodes + extra_unspt_nodes
 
 def _disc_engine_conversion(module):
-    def try_cvt_to_disc_engine_func(c_module, subgraph, group_name, q_info=None):
+    def try_cvt_to_disc_engine_func(
+            c_module, subgraph, group_name, q_info=None, grp_calib_data=None
```
Comment (fortianyou) [inline pytorch_blade/torch_blade/mlir/disc_engine_conversion.py:109]: Nit: grp_calib_data is not used for now here. Add it later when we support quantization in disc.

File: pytorch_blade/torch_blade/tensorrt/tensorrt_optimization.py [hunk 2]
```diff
@@ -80,4 +88,7 @@ def optimize_trt(script_module, q_info=None, disable_fallback=False):
     dynamic_settings = get_dynamic_settings(c_module, trt_unsupported)
     top_level_block = graph
     supported_node_fusion(graph, top_level_block, trt_unsupported, q_info)
-    trt_engine_conversion(c_module, q_info, disable_fallback, dynamic_settings)
+    with get_calib_file_for_each_group(c_module) as quantization_calib_file:
+        trt_engine_conversion(
+            c_module, q_info, disable_fallback, dynamic_settings, quantization_calib_file
+        )
```
Comment (fortianyou) [inline pytorch_blade/torch_blade/tensorrt/tensorrt_optimization.py:94]: remove quantization_calib_file when exiting the context.

File: pytorch_blade/torch_blade/quantization/prepare_data.py [hunk 3]
```diff
@@ -195,28 +196,28 @@ def get_calib_data_for_each_group(self):
...
-        logger.warning(f"Unable to get calib file for each graph group due to {e}.")
-        return None
+    if trt_calibration_data is not None:
+        try:
+            data_preparer = DataPreparer(c_module, trt_calibration_data)
+            calib_data_for_all_fusion_group = data_preparer.prepare()
+        except Exception as e:
+            logger.warning(f"Unable to get calib file for each graph group due to {e}.")
+
+    with tempfile.TemporaryDirectory(prefix="torch_blade") as tmp_dirname:
```
Comment (fortianyou) [inline pytorch_blade/torch_blade/quantization/prepare_data.py:217]: Nit: How about making the tmp_dir implicit?
```
with tempfile.TemporaryDirectory(prefix=".torch_blade") as tmp_dirname:
```

File: pytorch_blade/src/compiler/backends/engine_interface.h [hunk 4]
```diff
@@ -18,6 +18,10 @@
 namespace torch {
 namespace blade {
 namespace backends {
+
+typedef std::vector<at::Tensor> SingleCalibInputType;
```
[the remark this answers is not in the document]
Comment (chenbohua3) [reply pytorch_blade/src/compiler/backends/engine_interface.h]: done

File: pytorch_blade/src/compiler/backends/engine_interface.h [hunk 5]
```diff
@@ -29,6 +33,11 @@ struct EngineState {
   TORCH_BLADE_BACKENDS_DEFINE_FIELD(outputs, std::vector<TensorType>);
   TORCH_BLADE_BACKENDS_DEFINE_FIELD(extra_attrs, AttrType);
 
+  // As a medium for passing values from python to c++ and
+  // will not be serialized
+
+  TORCH_BLADE_BACKENDS_DEFINE_FIELD(grp_calib_data, AllCalibDataType);
```
[the remark this answers is not in the document]
Comment (chenbohua3) [reply pytorch_blade/src/compiler/backends/engine_interface.h]: done

File: pytorch_blade/torch_blade/mlir/disc_engine_conversion.py (hunk 1 above)
Comment (chenbohua3) [reply pytorch_blade/torch_blade/mlir/disc_engine_conversion.py:109]: disc and trt share the same `group_node_to_engine`. If `grp_calib_data` is not added, there will be a "argument number mismatch" problem

File: pytorch_blade/torch_blade/quantization/prepare_data.py (hunk 3 above)
Comment (chenbohua3) [reply pytorch_blade/torch_blade/quantization/prepare_data.py]: done

File: pytorch_blade/torch_blade/tensorrt/tensorrt_optimization.py (hunk 2 above)
Comment (chenbohua3) [reply pytorch_blade/torch_blade/tensorrt/tensorrt_optimization.py:94]: If leave the context of `with tempfile.TemporaryDirectory(prefix="torch_blade") as tmp_dirname:`, the temporary directory will be automatically removed:)

File: pytorch_blade/src/pybind.cpp [hunk 6]
```diff
@@ -104,6 +104,7 @@ void initModules<false>(py::module& m) {
       .def_readwrite("backend_name", &EngineState::backend_name)
       .def_readwrite("inputs", &EngineState::inputs)
       .def_readwrite("outputs", &EngineState::outputs)
+      .def_readwrite("grp_calib_data", &EngineState::calib_data)
```
[the remark this answers is not in the document]
Comment (chenbohua3) [reply pytorch_blade/src/pybind.cpp]: done

**Review by fortianyou — APPROVED** [review]
LGTM, Thanks

**Pull request merged**